src/hooks/useStable.ts

```diff
@@ -0,0 +1,74 @@
+
+const stable = Symbol("stable");
+
+/**
+ * A marker type to indicate a value *must* be stable.
+ *
+ * This is typically used for a large/complicated component like GridTable that knows it's
+ * going to render ~many children, and so internally uses `useMemo`s to optimize performance,
+ * and needs to indicate to callers which props should be stable.
+ *
+ * ```ts
+ * export interface GridTableProps {
+ *   // The use of Stable tells callers they must pass a stable array.
+ *   columns: Stable<Column[]>;
+ * }
+ * ```
+ *
+ * Note that we use a value type of `any` so that the `[stable]` key does not
+ * cause type mismatches on args that have index signatures.
+ */
+export type Stable<T> = T & { [stable]: any };
+
+/** Limits the dependencies of `useStable` to be themselves only stable values. */
+export type StableDep = Stable<unknown> | string | number | boolean | null | undefined;
+
+/**
+ * Allows creating "stable values", which is really just a dressed up `useMemo`.
+ *
+ * The only difference between `useStable` and `useMemo` is that `useStable` restricts its
+ * dependencies to themselves be stable values. This is similar to the eslint rule that
+ * catches a locally-declared `const array = [1, 2, 3]` can't be used as a dependency,
+ * because it will change on every render, but moves the check into the TypeScript type
+ * system where we can more robustly enforce it.
+ */
+export function useStable<T>(fn: () => T, deps: ReadonlyArray<StableDep> | undefined): Stable<T> {
+  // We don't need the `fn` factory to be a dep so disable eslint warning
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  return useMemo(fn, deps) as Stable<T>;
+}
+
+/**
+ * Marks a value as stable even though it's not.
+ *
+ * This should only be used for unit tests, or for pages/components that the programmer
+ * "just knows" are extremely-small/read-only, such that passing in a prop that breaks
+ * memoization is okay.
+ */
+export function pretendStable<T>(value: T): Stable<T> {
+  return value as Stable<T>;
+}
+
+/**
+ * Downgrades the given `Props` by removing the `Stable` wrapper from each key to ease migration.
+ *
+ * I.e. a component can have its primary props type have the correct `Stable` marked types,
+ * but then provide a separate "unsafe" fallback/transition component:
+ *
+ * ```ts
+ * export const MyComponentUnsafe: (props: UnstableProps<MyComponentProps>) => ReactElement = MyComponent as any;
+ * ```
+ */
+export type UnstableProps<Props> = {
+  [K in keyof Props]: Props[K] extends Stable<infer T> ? T : Props[K];
+};
+
+declare module "react" {
+  // Extend useMemo's return values to be automatically considered stable.
+  //
+  // With this extension, we could almost get away without our own `useStable` hook,
+  // but we can't use the extension mechanism to restrict useMemo's `deps` to be
+  // `ReadonlyArray<StableDep>`.
+  function useMemo<T>(factory: () => T, deps: ReadonlyArray<unknown> | undefined): Stable<T>;
```

Comment:
@bdow FWIW during demos, answering a question for @khalidwilliams , I'm kinda thinking we should back away from this change, which "upgrades" `useMemo` to produce stable values, and have `useStable` be the only one that can make stable values.

The reason is that, with the TS typing, I'm unable to "upgrade" `useMemo`'s dependency list to be _only_ stable values, and it is precisely this restriction that we'd need to have prevented the bug I fixed.

I.e. the bug I fixed, the `const columns = useMemo(...)` was "already useMemo-d" _but_ the `elevations` dependency was itself an unstable array:

https://github.com/homebound-team/internal-frontend/pull/4159/files?w=1#diff-eb71015b91b3c704dd3488a15c973047da7e637251f2d4a5b2f467c0861c73f9R146

So, really, the fix we need to be 100% bulletproof to performance is not just: a) the user uses `useStable` or `useMemo` for the rows/columns props, but also b) any _dependencies_ the user used for `useStable` or `useMemo` must themselves be stable.

And I cannot "fix" `useMemo` to have this restriction (it would also be a huge breaking change), so I think we'd have to just say "now all `const rows = useMemo(...)`s have to be `const rows = useStable(...)`s" and deal with the fallout.

I'm pretty convinced this is the right direction, even though it's a more disruptive change, b/c all of our `useMemo`s lines need to become `useStable`s--and also make sure their dependencies are stables, but that's the whole point, we need to have the stableness check be transitively applied to dependencies of our dependency.